packages/next/src/server-build.ts

```diff
@@ -1363,7 +1380,30 @@ export async function serverBuild({
       ...denormalizeNextDataRoute(),
 
       // /_next/data routes for getServerProps/getStaticProps pages
-      ...dataRoutes,
+      ...(isNextDataServerResolving
+        ? // when resolving data routes for middleware we need to include
+          // all dynamic routes including non-SSG/SSP so that the priority
+          // is correct
+          dynamicRoutes
+            .map(route => {
+              route = Object.assign({}, route);
+              route.src = path.join(
```

Comment:
Since these are routes, I wonder if it should use `path.posix.join()` instead so this always uses `/` instead of `\` on windows

Reply:
Updated in https://github.com/vercel/vercel/pull/8278/commits/9e084beb1f9bfe6456a4ad83aa758f0c8c91f87b seems there are other cases we are using this so can update those in a follow-up PR.